core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,343 @@
+   * manager opens a new input stream.
+   *
+   * @param ufs the under file system
+   * @param path the path to the under storage file
+   * @param fileId the file id
+   * @param openOptions the open options
+   * @param reuse true to reuse existing input stream, otherwise acquire a new stream
+   * @return the acquired input stream
+   * @throws IOException if the input stream fails to open
+   */
+  public InputStream acquire(UnderFileSystem ufs, String path, long fileId, OpenOptions openOptions,
+      boolean reuse) throws IOException {
+    if (!ufs.isSeekable() || !isCachingEnabled()) {
+      // not able to cache, always return a new input stream
+      return ufs.open(path, openOptions);
+    }
+
+    // explicit cache cleanup
+    mUnderFileInputStreamCache.cleanUp();
+
+    UfsInputStreamIds resources;
+    synchronized (mFileIdToInputStreamIds) {
+      if (mFileIdToInputStreamIds.containsKey(fileId)) {
+        resources = mFileIdToInputStreamIds.get(fileId);
+      } else {
+        resources = new UfsInputStreamIds();
+        mFileIdToInputStreamIds.put(fileId, resources);
+      }
+    }
+
+    synchronized (resources) {
+      long nextId = UNAVAILABLE_RESOURCE_ID;
+      CachedSeekableInputStream inputStream = null;
+      if (reuse) {
+        // find the next available input stream from the cache
+        for (long id : resources.availableIds()) {
+          inputStream = mUnderFileInputStreamCache.getIfPresent(id);
+          if (inputStream != null) {
+            nextId = id;
+            LOG.debug("Reused the under file input stream resource of {}", nextId);
+            // for the cached ufs instream, seek to the requested position
+            inputStream.seek(openOptions.getOffset());
+            break;
+          }
+        }
+      }
+      // no cached input stream is available, open a new one
+      if (nextId == UNAVAILABLE_RESOURCE_ID) {
+        nextId = IdUtils.getRandomNonNegativeLong();
+        final long newId = nextId;
+        try {
+          inputStream = mUnderFileInputStreamCache.get(nextId, () -> {
+            SeekableUnderFileInputStream ufsStream = (SeekableUnderFileInputStream) ufs.open(path,
+                OpenOptions.defaults().setOffset(openOptions.getOffset()));
+            LOG.debug("Created the under file input stream resource of {}", newId);
+            return new CachedSeekableInputStream(ufsStream, newId, fileId, path);
+          });
+        } catch (ExecutionException e) {
+          LOG.warn("Failed to create a new cached ufs instream of file id {} and path {}", fileId,
+              path);
+          // fall back to a ufs creation.
+          return ufs.open(path, OpenOptions.defaults().setOffset(openOptions.getOffset()));
+        }
+      }
+
+      // mark the input stream id as acquired
+      resources.acquire(nextId);
+      return inputStream;
+    }
+  }
+
+  private boolean isCachingEnabled() {
```

Comment:
why not defining a CONST instead of having a helper method?

Reply:
sgtm